### PR TITLE
Add unit tests for syncQueryParamsWithURL

### PR DESCRIPTION
**Pull request by gusaiani**
Follow-up to #16795 

To run tests locally and see coverage: 

```sh
yarn test-unit frontend/test/metabase/parameters/components/Parameters/syncQueryParamsWithURL.unit.spec.js --coverage
```

You will see a line stating coverage at ~91% for the implementation file.

Getting it to 100% will be a mix of:

1. treating the underlying prop architecture in the implementation, or throwing if no `setParameterValue` prop is passed
2. adding coverage to edge cases in the parsing helper function

These started looking like low ROI given the broader context, so I am submitting for review as is.

image

### CODE REVIEW

File: frontend/test/metabase/parameters/components/Parameters/syncQueryParamsWithURL.unit.spec.js [hunk 1]
```diff
@@ -2,9 +2,215 @@ import Field from "metabase-lib/lib/metadata/Field";
...
+          expect(setParameterValue).toHaveBeenCalledTimes(1);
+        });
+
+        it("calls setParameterValue with one object as argument, keys of parameter id and parsed param values", () => {
+          syncQueryParamsWithURL(props);
+          expect(setParameterValue).toHaveBeenCalledWith({
+            id1: "2021",
+            id2: "CA",
+          });
+        });
```
Comment (kulyk) [inline frontend/test/metabase/parameters/components/Parameters/syncQueryParamsWithURL.unit.spec.js:209]: I guess we can improve the test names here. Right now they are mostly identical to the assert statements, so they don't bring new context about the behavior we test, and a developer needs to spend some extra effort to figure out what's happening here.

I believe test names should explain _why_ the behavior should be like that, and _how_ usually visible from the code itself. Also, unit tests become much more valuable in this case as they also become some sort of a spec or doc for a function.

E.g. it can be something like that (I'm sure the wording can be better, but I hope it still represents the idea):

```js
it("doesn't change anything if there are no matching parameters", () => {
  const props = {
    parameters: [
      {
        id: "some-param",
        slug: "noMatchingQueryParam",
      },
    ],
    query: {
      someQueryParam: "noMatchingParameter",
    },
  };

  syncQueryParamsWithURL(props);

  expect(setParameterValue).not.toHaveBeenCalled();
});

it("syncs parameter with query params by slugs", () => {
  const props = {
    parameters: [
      {
        id: "created-at-param-id",
        slug: "createdAt",
      },
    ],
    query: {
      createdAt: "2021",
    },
  };

  syncQueryParamsWithURL(props);

  expect(setParameterValue).toHaveBeenCalledTimes(1);
  expect(setParameterValue).toHaveBeenCalledWith({
    "created-at-param-id": "2021",
  });
});

it("syncs multiple parameters with query params by slugs", () => {
  const props = {
    parameters: [
      {
        id: "created-at-param-id",
        slug: "createdAt",
      },
      {
        id: "state-param-id",
        slug: "state",
      },
    ],
    query: {
      createdAt: "2021",
      state: "CA",
    },
  };

  syncQueryParamsWithURL(props);

  expect(setParameterValue).toHaveBeenCalledTimes(1);
  expect(setParameterValue).toHaveBeenCalledWith({
    "created-at-param-id": "2021",
    "state-param-id": "CA",
  });
});
```

File: frontend/test/metabase/parameters/components/Parameters/syncQueryParamsWithURL.unit.spec.js [hunk 2]
```diff
@@ -2,9 +2,215 @@ import Field from "metabase-lib/lib/metadata/Field";
...
+    describe("for internal questions", () => {
+      describe("when parameters length is 0", () => {
+        const props = {
+          commitImmediately: true,
+          parameters: [],
+          query: {
+            createdAt: "2021",
+          },
+          setParameterValue,
+        };
```
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/test/metabase/parameters/components/Parameters/syncQueryParamsWithURL.unit.spec.js:27]: Great, if that makes the test more understandable to more people, I'm in.

File: frontend/test/metabase/parameters/components/Parameters/syncQueryParamsWithURL.unit.spec.js (hunk 1 above)
Comment (gusaiani) [reply frontend/test/metabase/parameters/components/Parameters/syncQueryParamsWithURL.unit.spec.js:209]: Yup, good call, will update 🙇

Comment (gusaiani) [reply frontend/test/metabase/parameters/components/Parameters/syncQueryParamsWithURL.unit.spec.js]: @kulyk, thank you for your review.

I've just updated the PR based on your feedback.

File: frontend/test/metabase/parameters/components/Parameters/syncQueryParamsWithURL.unit.spec.js [hunk 3]
```diff
@@ -2,9 +2,210 @@ import Field from "metabase-lib/lib/metadata/Field";
...
+              slug: "state",
+            },
+          ],
+          query: {
+            createdAt: "2021",
+            state: "CA",
+          },
+        });
+
+        it("syncs parameter with query params by slugs, by passing one object with as many key/value pairs as there are matching parameters, all in a single function call", () => {
```
Comment (kulyk) [inline frontend/test/metabase/parameters/components/Parameters/syncQueryParamsWithURL.unit.spec.js:197]: Maybe too verbose at my taste, but we better hear another opinion here :D

**Review by kulyk — APPROVED** [review]
LGTM! Thank you @gusaiani 👏
